packages/financial-templates-lib/src/price-feed/QuandlPriceFeed.js

```diff
@@ -0,0 +1,248 @@
+
+// An implementation of PriceFeedInterface that uses the Quandl free API to retrieve prices.
+// API details can be found here: https://docs.quandl.com/docs
+class QuandlPriceFeed extends PriceFeedInterface {
+  /**
+   * @notice Constructs the QuandlPriceFeed.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} web3 Provider from truffle instance to connect to Ethereum network.
+   * @param {String} quandlApiKey Quandl Data API key. Will get rate-limited without an API key.
+   * @param {String} datasetCode Code identifying the database to which the dataset belongs. Example: "CHRIS".
+   * @param {String} databaseCode Code identifying the dataset. Example: "CME_MGC1".
+   * @param {Integer} lookback How far in the past the historical prices will be available using getHistoricalPrice.
+   * @param {Object} networker Used to send the API requests.
+   * @param {Function} getTime Returns the current time.
+   * @param {Number} priceFeedDecimals Number of priceFeedDecimals to use to convert price to wei.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   */
+  constructor(
+    logger,
+    web3,
+    quandlApiKey,
+    datasetCode,
+    databaseCode,
+    lookback,
+    networker,
+    getTime,
+    priceFeedDecimals = 18,
+    minTimeBetweenUpdates = 43200
+    // 12 hours is a reasonable default since this pricefeed returns daily granularity at best.
+  ) {
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+
+    this.apiKey = quandlApiKey;
+    this.datasetCode = datasetCode.toUpperCase();
+    this.databaseCode = databaseCode.toUpperCase();
+    this.lookback = lookback;
+    this.uuid = `Quandl-${datasetCode}-${databaseCode}`;
+    this.networker = networker;
+    this.getTime = getTime;
+    this.minTimeBetweenUpdates = minTimeBetweenUpdates;
+    this.priceFeedDecimals = priceFeedDecimals;
+
+    this.toBN = this.web3.utils.toBN;
+
+    this.convertPriceFeedDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number.
+      // Note: Must ensure that `number` has no more decimal places than `priceFeedDecimals`.
+      return this.toBN(parseFixed(number.toString().substring(0, priceFeedDecimals), priceFeedDecimals).toString());
+    };
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  async getHistoricalPrice(time, verbose = false) {
+    if (this.lastUpdateTime === undefined) {
+      throw new Error(`${this.uuid}: undefined lastUpdateTime`);
+    }
+
+    // Set first price period in `historicalPricePeriods` to first non-null price.
+    let firstPricePeriod;
+    for (let p in this.historicalPricePeriods) {
+      if (this.historicalPricePeriods[p] && this.historicalPricePeriods[p].openTime) {
+        firstPricePeriod = this.historicalPricePeriods[p];
```

Comment:
is this just finding the first non empty instance or something?

Reply:
Yup, copied from the CryptoWatch code. It finds the first price period (assuming they are sorted oldest first), in order to check that the request timestamp >= firstPricePeriod.openTime, otherwise the timestamp is before all data and we can exit early. Also if there is no `firstPricePeriod` after this loop, then we can also exit early because there is no data.